add ebs volume info to the cluster report

gorskysd opened the pull request:
Add ebs volume information to the cluster report. Information is gathered via the `describe_volumes` call. This is necessary because the `describe_instances` call does not provide comprehensive information about any attached ebs storage, including the storage type, IOPS, and throughput.

To enable reusing most of the existing code without refactor I added the list of volumes as an attribute in the `instances` dict. I considered break this dict apart in the cluster report for a sense of "separation of duties", but I wasn't sure if that was any better. For example:
```
class AWSDatabricksClusterReport(DatabricksClusterReport):
    instances: Union[dict, None]
    volumes: Union[dict, None] #  <--- new field for volumes?
```

gorskysd commented on sync/awsdatabricks.py:
```diff
@@ -308,7 +307,7 @@ def monitor_cluster(cluster_id: str, polling_period: int = 30) -> None:
 
 
-def _monitor_cluster(
+def _monitor_cluster(  # noqa: C901
```
Obviously not ideal, but this addition barely increased the complexity over the line, and I didn't see any obvious factoring that would *actually* make this logic clearer to understand.

gorskysd commented on sync/awsdatabricks.py:
```diff
@@ -388,10 +381,69 @@ def write_file(body: bytes):
+                {"Name": "tag:ClusterId", "Values": [cluster_id]},
+            ],
+            **args,
+        )
+
+        if new_volumes := response.get("Volumes"):
+            volumes += new_volumes
+
+        if next_token := response.get("NextToken"):
+            args["NextToken"] = next_token
+        else:
+            break
+
+    return volumes
+
+
+def _get_ec2_instances(cluster_id: str, ec2_client: "botocore.client.ec2") -> dict[str, list]:
+
+    args = {}
+    instances = []
+    while True:
+        response = ec2_client.describe_instances(
+            Filters=[
+                {"Name": "tag:Vendor", "Values": ["Databricks"]},
+                {"Name": "tag:ClusterId", "Values": [cluster_id]},
+            ],
+            **args,
+        )
+
+        print(response)
+
+        if new_instances := response.get("Reservations"):
+            instances += new_instances
+
+        if next_token := response.get("NextToken"):
+            args["NextToken"] = next_token
+        else:
+            break
+
+    return {"Reservations": instances}
```
The `describe_volumes` call is new. I also factored out the `describe_instances` call because it was common to two places, and it was a good opportunity to beef it up with some NextToken handling for very large clusters.

rmoneys commented on sync/awsdatabricks.py:
```diff
@@ -388,10 +382,67 @@ def write_file(body: bytes):
+                    recorded_volume_id_set.add(v["VolumeId"])
+                    recorded_volumes.append(v)
 
+            instances["Volumes"] = recorded_volumes
```
This solution won't extend to Azure Databricks because the instances in a cluster report is an array instead of a dict: https://github.com/synccomputingcode/syncsparkpy/blob/c5575c873acc150e9f5d931f0adca0c8296d2122/sync/models.py#L110-L111
We could make it a dict, but seems better to put volumes under "volumes" at the top level if only for better naming.

gorskysd replied:
Gotcha, that's something I was considering doing, and I think I like it more. It maintains some parity with Azure (though idk if a separate call like this is needed for Azure), but it also keeps the existing structures unchanged which has some benefit to mental load.

rmoneys replied:
You said it better than me!

rmoneys commented on sync/awsdatabricks.py:
```diff
@@ -388,10 +382,67 @@ def write_file(body: bytes):
+            for v in volumes:
+                if v["VolumeId"] not in recorded_volume_id_set:
+                    recorded_volume_id_set.add(v["VolumeId"])
+                    recorded_volumes.append(v)
```
Hopefully that'll remove the need for the noqa directive

gorskysd replied:
Good call, noqa is gone

gseyffert-sync commented on sync/awsdatabricks.py:
```diff
@@ -388,10 +382,67 @@ def write_file(body: bytes):
+
+    args = {}
+    volumes = []
+    while True:
```
For both this function and `_get_ec2_instances`, I personally prefer a pattern more like -

```python
response = ec2_client.describe_volumes(...)
volumes = response.get("Volumes", [])
next_token = response.get("NextToken")

while next_token:
    response = ec2_client.describe_volumes(..., NextToken=next_token)
    volumes += response.get("Volumes", [])
    next_token = response.get("NextToken")

return volumes
```

This eliminates the need for a `break` in your loop, and explicitly separates the initial load logic from the paging logic

gseyffert-sync replied:
I will add that I used to always try to put the initial / paging logic in the same loop, but I've given up on that. Separating out the two now generally, I find, leads to code that is easier to follow / update

gorskysd replied:
Happy to make this change. Just curious, does loop breaking like this make you nervous from a brittleness standpoint?

gseyffert-sync commented on sync/awsdatabricks.py:
```diff
@@ -1,8 +1,10 @@
 import logging
 from time import sleep
+from typing import Dict, List
```
Noticed you initially were using `dict`/`list` - just wanted to explicitly state, in case you were like "wft?", that this is for Python 3.7 support to be compatible with older Databricks versions :+1:

gorskysd replied:
Yeeeep. I remember when we upgraded from 3.7 to 3.10. I knew what was up right away.

gorskysd commented:
Thanks for the feedback, y'all. In the process of making "volumes" a higher tier entity I broke some tests. Fixed those, and added a mock for the `describe_volumes` call.

I tested the changes on both an EBS cluster and a storage attached NVMe cluster (though the latter still has two ebs volumes attached per instance by default).

gseyffert-sync commented on sync/awsdatabricks.py:
```diff
@@ -216,7 +221,7 @@ def _get_cluster_report(
 
 
-def _get_cluster_instances(cluster: dict) -> Response[dict]:
+def _get_cluster_instances(cluster: dict) -> Tuple[Response[dict], Response[dict]]:
```
nit - Perhaps worth re-naming this function to something like `_get_cluster_aws_information`?

gseyffert-sync reviewed (approved):
I think this LGTM. Just to make sure I read it correctly, the top-level file structure that we write in `monitor_cluster` will be something like -

```javascript
{
  "Reservations": [...],
  "Volumes": [...],
}
```

Where `"Reservations"`/`"Volumes"`` are both top-level keys from the corresponding AWS `describe_*` calls?

Since we are doing this for more than just instances now, we can probably remove some of the intermediate `"Reservations"` usages that are sprinkled around here, but I can also handle that in some upcoming work I need to do

gorskysd commented:
@gseyffert-sync on second thought, those nits are rather important for clarity down the line. 

I did a whole bunch of relabeling to avoid conflating response data from `describe_instances` ("Reservations") and `describe_volumes` ("Volumes"). These two objects do get combined in the file and in the cluster report as:
```
# In saved file
aws_cluster_info = {
    "Reservations" : ...
    "Volumes" : ...
}

# In cluster report
AWSDatabricksClusterReport(
    instances : {"Reservations" : [...]}
    volumes: {"Volumes" : [...]}
)
```

A few other things that changed:

- (!!) Changed the stored filename to `aws_cluster_info.json`. This was for consistency points. If you think it will be problematic I'll change it back!
- A few methods had their definitions changed to use `aws_cluster_info`. EXCEPT for `_get_cluster_instances_from_dbfs` imported from `_databricks.py`.
- Made `_get_ec2_instances` and `_get_ebs_volumes` symmetric such that they both remove the top-tier field and just return the list instead. This allowed for a lot of simplification in `_monitor_cluster`. 
    * The top-tier fields are reintroduced when saving the file, and both are persisted in the cluster report as well. 
    * Obviously the double nestings in the cluster report are unnecessary, but I'd rather treat these two AWS entities symmetrically throughout the library, and also not change what the api expects for `instances`.

gseyffert-sync commented on sync/_databricks.py:
```diff
@@ -223,7 +223,7 @@ def _get_cluster_report(
+def _get_aws_cluster_info_from_dbfs(filepath: str):
     filepath = format_dbfs_filepath(filepath)
     dbx_client = get_default_client()
     try:
```
sorry, I tried to cancel my review - these comments are no longer relevant

gseyffert-sync commented:
Looks good to me! We will need to add this `volumes` field to the API model as well, I'd imagine

rmoneys commented on sync/awsdatabricks.py:
```diff
@@ -227,64 +232,67 @@ def _get_cluster_instances(cluster: dict) -> Response[dict]:
             logger.error(
-                f"Unexpected error encountered while attempting to fetch sync_data/cluster_instances.json: {ex}"
+                f"Unexpected error encountered while attempting to fetch sync_data/aws_cluster_info.json: {ex}"
             )
```
Potential improvement:
```
except ClientError as err:
    logger.warning(f"Failed to retrieve cluster info from S3 with key, '{file_key}': {err}")
```
The reason for the exception is not important to the consumer which anticipates a `None` response so it should be a warning in either case. If we need to better understand the reason it'll be in the log statement, e.g. "Failed to retrieve cluster info from S3 with key, 'sync_data/1234210239561239/aws_cluster_info.json': An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist."

gorskysd replied:
Done! Happy to make changes that cut out lines :D

rmoneys reviewed (commented):
Looks good, but our web API won't see volume info until we make some changes further up the call hierarchy. sync._databricks.create_prediction_for_run ignores the "volumes" attribute on the AWSDatabricksClusterReport - do we want use volume info in the current prediction model?

gorskysd commented:
Originally wasn't gonna connect this to the API since it isn't really needed with the current predictor version, but I might as well just do it and make the backend changes now. I'll spin up a sister branch for the backend.

rmoneys reviewed (dismissed):
You may also want to update the version in sync.__info__.py if you want to make this available to users, but approving in case you think this is a good unit of work as is.
